xarray/backends/api.py

```diff
@@ -1023,8 +1027,24 @@ def save_mfdataset(datasets, paths, mode='w', format=None, groups=None,
                              for w, s in zip(writes, stores)])
 
 
+def _validate_datatypes_for_zarr_append(dataset):
+    """DataArray.name and Dataset keys must be a string or None"""
+    def check_dtype(var):
+        if (not np.issubdtype(var.dtype, np.number)
+                and not np.issubdtype(var.dtype, np.string_)
+                and not np.issubdtype(var.dtype, np.unicode_)
+                and not var.dtype == object):
```

Comment:
```suggestion
                and not coding.strings.is_unicode_dtype(var.dtype)):
```

Reply:
this seems to work:
```python
         if (not np.issubdtype(var.dtype, np.number)
                and not coding.strings.is_unicode_dtype(var.dtype)
                and not var.dtype == object):
```
while this does not:
```python
         if (not np.issubdtype(var.dtype, np.number)
                and not np.issubdtype(var.dtype, np.string_)
                and not np.issubdtype(var.dtype, np.unicode_)
                and not coding.strings.is_unicode_dtype(var.dtype)):
```

Reply:
or were you suggesting this?
```python
         if (not np.issubdtype(var.dtype, np.number)
                and not coding.strings.is_unicode_dtype(var.dtype)):
```
this does not work

Reply:
OK, I understand why this doesn't work: we aren't using xarray's standard encoding pipeline when appending variables. This probably is fine for strings -- in the worst case, users will just get an error downstream from Zarr instead.